File: TechElevatePlan/Codes/Has_Duplicate.py
def hasDuplicateValue(array):
    existingNumbers = []
    for i in range(len(array)):
        if existingNumbers[array[i]] == 1:
            return True
        else:
            existingNumbers[array[i]] = 1
    return False


"""Time Complexity - O(N)
   Space Complexity - O(N)
"""
def has_duplicate_value(array):
    # create an empty dictionary to store existing values
    existing_values = {}
    # loop through the array using a for loop
    for i in range(len(array)):
        # if the current value in the array does not exist in the dictionary
        if not existing_values.get(array[i]):
            # add the current value to the dictionary with a value of True
            existing_values[array[i]] = True
        else:
            # if the current value already exists in the dictionary, return True
            return True
    # if no duplicates are found, return False
    return False

"""Time Complexity - O(logN)
   Space Complexity - O(logN)
"""
def has_duplicate_value(array):
    # sort the array in ascending order
    array.sort()
    # loop through the sorted array
    for i in range(len(array) - 1):
        # if the current element is equal to the next element
        if array[i] == array[i + 1]:
            # return True as a duplicate value is found
            return True
    # if no duplicates are found, return False
    return False

File: TechElevatePlan/Codes/Has_Duplicate.js

"Quadratic Time Complexity O(n^2)"
function hasDuplicateValue(array) {
    let steps = 0; // count of steps
    for(let i = 0; i < array.length; i++) {
        for(let j = 0; j < array.length; j++) {
            steps++; // increment number of steps
            if(i !== j && array[i] === array[j]) {
                return true;
            }
        }
    }
    console.log(steps); // print number of steps if no duplicates
    return false;
}

"Linear Time Complexity O(n)"
function hasDuplicateValue(array) {
    // create an empty array to store the numbers we've seen
    let existingNumbers = [];
    // Use a loop to check each number in the array
    for(let i = 0; i < array.length; i++) {
        // if the number is already in the array, return true
        // === 1 is used to check if the number is in the array
        // As it encounters each
        // number, it places an arbitrary value (we’ve chosen to use a 1) in the existingNumbers
        // array at the index of the number we’re encountering.
        if(existingNumbers[array[i]] === 1) {
            return true;
        } else {
            existingNumbers[array[i]] = 1;
        }
    }
    return false;
}